### PR TITLE
Add ScatterFrom property to AbsorptionCorrection

### PR DESCRIPTION
**Pull request by peterfpeterson**
This is another step in getting absorption correction calculated for every sample for total scattering. The intended workflow (like in the test), is to `SetSample` then run `AbsorptionCorrection` multiple times to get the correction for each sub-component in the beam (sample, container, environment).

There is some amount of cleanup where `Rasterize` was changed to work off of `const &` rather than pointers. This is an attempt to write more sensible code.

**To test:**

A unit test was added to exercise the functionality. Reviewing the code is more useful.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/AbsorptionCorrection.cpp [hunk 1]
```diff
@@ -78,6 +83,15 @@ void AbsorptionCorrection::init() {
                                                    Direction::Output),
                   "Output workspace name");
 
+  // AbsorbedBy
+  std::vector<std::string> scatter_options{CALC_SAMPLE, CALC_CONTAINER,
+                                           CALC_ENVIRONMENT};
+  declareProperty(
+      "ScatterFrom", CALC_SAMPLE,
+      boost::make_shared<StringListValidator>(scatter_options),
+      "Select the method to use to calculate exponentials, normal or a\n"
```
Comment (martyngigg) [inline Framework/Algorithms/src/AbsorptionCorrection.cpp]: I think this comment needs updating?

File: Framework/Algorithms/src/AbsorptionCorrection.cpp [hunk 2]
```diff
@@ -78,6 +83,15 @@ void AbsorptionCorrection::init() {
                                                    Direction::Output),
                   "Output workspace name");
 
+  // AbsorbedBy
+  std::vector<std::string> scatter_options{CALC_SAMPLE, CALC_CONTAINER,
+                                           CALC_ENVIRONMENT};
+  declareProperty(
+      "ScatterFrom", CALC_SAMPLE,
+      boost::make_shared<StringListValidator>(scatter_options),
```
Comment (martyngigg) [inline Framework/Algorithms/src/AbsorptionCorrection.cpp]: `scatter_options` could me moved in I think?

File: Framework/Algorithms/src/AbsorptionCorrection.cpp [hunk 3]
```diff
@@ -323,17 +385,30 @@ void AbsorptionCorrection::retrieveBaseProperties() {
...
-    m_sampleObject = &sample.getShape();
+    // Get the shape from the proper object
+    if (scatterFrom == CALC_SAMPLE)
+      m_sampleObject = &sample.getShape();
+    else if (scatterFrom == CALC_CONTAINER)
+      m_sampleObject = &(sample.getEnvironment().getContainer());
+    else if (scatterFrom == CALC_ENVIRONMENT)
+      m_sampleObject = &(sample.getEnvironment().getComponent(1));
+    else
+      throw std::runtime_error("Somebody forgot to fill in an if/else tree");
```
Comment (martyngigg) [inline Framework/Algorithms/src/AbsorptionCorrection.cpp]: Maybe use an `assert` here instead as it logically shouldn't get here?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Generally looks good. A couple of minor comments so far.

File: Framework/Algorithms/src/AbsorptionCorrection.cpp (hunk 3 above)
Comment (peterfpeterson) [reply Framework/Algorithms/src/AbsorptionCorrection.cpp]: `assert` only functions in debug builds. I was trying to give a clear (enough) error message in the off chance that a user encountered it.

Comment (martyngigg) [reply Framework/Algorithms/src/AbsorptionCorrection.cpp]: Ok. I think maybe a slightly more clear help message at least indicating the parameter that is unexpected?

File: Framework/Geometry/inc/MantidGeometry/Rasterize.h [hunk 4]
```diff
@@ -32,30 +32,16 @@ struct MANTID_GEOMETRY_DLL Raster {
 namespace Rasterize {
 
 MANTID_GEOMETRY_DLL Raster calculate(const Kernel::V3D &beamDirection,
-                                     const Geometry::CSGObject &shape,
+                                     const IObject &shape,
```
Comment (martyngigg) [inline Framework/Geometry/inc/MantidGeometry/Rasterize.h]: :+1: for switching to just a reference-based implementation

**Review by martyngigg — APPROVED** [review]
Looks good once we've got the clearer help message and the merge conflict has been sorted.

**Pull request merged**